ouroboros-consensus-cardano-tools/src/Cardano/Tools/ImmDBServer/Diffusion.hs

```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE BlockArguments      #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE NamedFieldPuns      #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Cardano.Tools.ImmDBServer.Diffusion (run) where
+
+import           Control.Monad.Trans.Class
+import           Control.Monad.Trans.Cont
+import           Control.Tracer
+import qualified Data.ByteString.Lazy as BL
+import           Data.Functor.Contravariant ((>$<))
+import           Data.Void (Void)
+import           Network.Socket (SockAddr (..))
+
+import           Ouroboros.Network.ErrorPolicy (nullErrorPolicies)
+import           Ouroboros.Network.IOManager (withIOManager)
+import           Ouroboros.Network.Mux
+import qualified Ouroboros.Network.NodeToNode as N2N
+import qualified Ouroboros.Network.Snocket as Snocket
+import           Ouroboros.Network.Socket (configureSocket)
+
+import           Cardano.Tools.ImmDBServer.MiniProtocols (immDBServer)
+import           Ouroboros.Consensus.Block
+import           Ouroboros.Consensus.Config
+import           Ouroboros.Consensus.Config.SupportsNode
+import           Ouroboros.Consensus.Node.InitStorage
+                     (NodeInitStorage (nodeCheckIntegrity, nodeImmutableDbChunkInfo))
+import           Ouroboros.Consensus.Node.NetworkProtocolVersion
+import           Ouroboros.Consensus.Node.Run (SerialiseNodeToNodeConstraints)
+import           Ouroboros.Consensus.Storage.FS.API (SomeHasFS (..))
+import           Ouroboros.Consensus.Storage.FS.API.Types
+                     (MountPoint (MountPoint))
+import           Ouroboros.Consensus.Storage.FS.IO (ioHasFS)
+import qualified Ouroboros.Consensus.Storage.ImmutableDB as ImmutableDB
+import           Ouroboros.Consensus.Util
+import           Ouroboros.Consensus.Util.IOLike
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+serve ::
+     SockAddr
+  -> N2N.Versions N2N.NodeToNodeVersion N2N.NodeToNodeVersionData
+       (OuroborosApplication 'ResponderMode SockAddr BL.ByteString IO Void ())
+  -> IO Void
+serve sockAddr application = evalContT do
```

Comment:
Suggestion: Do not use `ContT`, use direct "bracket-style". I have not seen`ContT` used anywhere else in the codebase, I don't have a strong opinion against it but direct function calls seem simpler.

Reply:
Yeah, I think `ContT` "scales" better when you have lots of nested `bracket`-like things; but here, we don't have to reach for that; will remove :+1: